lib/iris/common/metadata.py

```diff
@@ -89,25 +98,428 @@ class BaseMetadata(metaclass=_NamedTupleMeta):
+                # is a hashable key.
+                args, kwargs = (), {qualname(func_service): False}
+
+            with LENIENT.context(*args, **kwargs):
+                result = func_operation(other)
+
+        return result
+
+    def _combine(self, other):
+        """Perform associated metadata member combination."""
+        if LENIENT(self.combine):
+            # Perform "lenient" combine.
+            logger.debug("lenient", extra=dict(cls=self.__class__.__name__))
+            values = self._combine_lenient(other)
+        else:
+            # Perform "strict" combine.
+            logger.debug("strict", extra=dict(cls=self.__class__.__name__))
+
+            def func(field):
+                value = getattr(self, field)
+                return value if value == getattr(other, field) else None
+
+            # Note that, for strict we use "_fields" not "_members".
+            values = [func(field) for field in self._fields]
+
+        return values
+
+    def _combine_lenient(self, other):
+        """
+        Perform lenient combination of metadata members.
+
+        Args:
+
+        * other (BaseMetadata):
+            The other metadata participating in the lenient combination.
+
+        Returns:
+            A list of combined metadata member values.
+
+        """
+
+        def func(field):
+            left = getattr(self, field)
+            right = getattr(other, field)
+            result = None
+            if field == "units":
+                # Perform "strict" combination for "units".
+                result = left if left == right else None
+            elif self._is_attributes(field, left, right):
+                result = self._combine_lenient_attributes(left, right)
+            else:
+                if left == right:
+                    result = left
+                elif left is None:
+                    result = right
+                elif right is None:
+                    result = left
+            return result
+
+        # Note that, we use "_members" not "_fields".
+        return [func(field) for field in BaseMetadata._members]
+
+    @staticmethod
+    def _combine_lenient_attributes(left, right):
+        """Leniently combine the dictionary members together."""
+        sleft = set(left.items())
+        sright = set(right.items())
+        # Intersection of common items.
+        common = sleft & sright
+        # Items in sleft different from sright.
+        dsleft = dict(sleft - sright)
+        # Items in sright different from sleft.
+        dsright = dict(sright - sleft)
+        # Intersection of common item keys with different values.
+        keys = set(dsleft.keys()) & set(dsright.keys())
+        # Remove (in-place) common item keys with different values.
+        [dsleft.pop(key) for key in keys]
+        [dsright.pop(key) for key in keys]
+        # Now bring the result together.
+        result = dict(common)
+        result.update(dsleft)
+        result.update(dsright)
```

Comment:
Yup, I'd considered this... but they've explicitly set the `history` key to be `None`, which might be meaningful to them, and in that case the `history` values are indeed actually different and can't be combined.

So I decided this was a step too far, and maybe a little unexpected for the user i.e., I've not applied leniency down to that microscopic level. It's completely doable, and we need to decide if that's what should happen out the box. It wasn't totally clear to me, so I erred on the side of it being a corner case and didn't support it.

Happy to discuss further, but I can't ever recall seeing a `None` value in any user metadata `attributes`... doesn't mean that this isn't a use case though.

Reply:
Okay I can agree to this behaviour, particularly due to this point
> but they've explicitly set the history key to be None, which might be meaningful to them, and in that case the history values are indeed actually different and can't be combined.
If we automatically created attributes for the user that defaulted to None, then I think this could trip user's up - they don't they have an attribute with None - but if the user has to create the attribue themselves then they shouldn't be caught out.